### PR TITLE
feat(sort) make orderBy work with locality by using Intl.collator | give…

### LINKED ISSUE #562: make orderBy work with locality by using Intl.collator
## Description

We want to try to make orderBy work with locality by using Intl.collator so that js-data can sort the languages with accents correctly. 

We also want to add a custom comparer callback function in-case we need to compare some special models outside of the property scope

## Environment

- js-data version: latest@master
- node or browser version: all browsers
- operating system:

## Steps to reproduce

try sorting Thai language in the current version, it wont sort correctly

```
{ test: 'คลอน' },
{ test: 'กลอน' },
{ test: 'สาระ' },
{ test: 'ศาลา' },
{ test: 'จักรพรรณ' },
{ test: 'จักรพรรดิ' },
{ test: 'เก๋ง' },
{ test: 'เก้ง' },
{ test: 'เก็ง' },
{ test: 'เก่ง' }
```
Should equal
```
{ test: 'กลอน' },
{ test: 'เก็ง' },
{ test: 'เก่ง' }
{ test: 'เก้ง' },
{ test: 'เก๋ง' },
{ test: 'คลอน' },  
{ test: 'สาระ' },
{ test: 'จักรพรรณ' },
{ test: 'จักรพรรดิ' },
{ test: 'ศาลา' },
{ test: 'สาระ' },
```

```
const store = new JSData.DataStore()
    store.defineMapper('item')
    const items = [
      { test: 'คลอน' },
      { test: 'กลอน' },
      { test: 'สาระ' },
      { test: 'ศาลา' },
      { test: 'จักรพรรณ' },
      { test: 'จักรพรรดิ' },
      { test: 'เก๋ง' },
      { test: 'เก้ง' },
      { test: 'เก็ง' },
      { test: 'เก่ง' }
    ]
    store.add('item', items)
    const params = {}

    params.orderBy = [
      ['test', 'ASC']
    ]
    assert.objectsEqual(store.query('item').filter(params).run(), [
      items[1],
      items[8],
      items[9],
      items[7],
      items[6],
      items[0],
      items[4],
      items[5],
      items[3],
      items[2]
    ])
```

### PR DESCRIPTION
**Pull request by wickstargazer**
… an option for custom compare if def[0] is a function

Fixes #562 (it's a good idea to open an issue first for discussion)

- [x] - `npm test` succeeds
- [ ] - Code coverage does not decrease (if any source code was changed)
- [ ] - If applicable, appropriate JSDoc comments were updated in source code (if applicable)
- [ ] - If applicable, approprate changes to js-data.io docs have been suggested ("Suggest Edits" button)

### CODE REVIEW

File: src/Query.js [hunk 1]
```diff
@@ -263,13 +263,21 @@ export default Component.extend({
...
-    if (cA < cB) {
-      return -1
-    } else if (cA > cB) {
-      return 1
+    /* Fix: compare by using collator */
+    let isNumeric = false
+    if (utils.isNumber(cA) || utils.isNumber(cB)) {
+      isNumeric = true
+    }
+    const collator = new Intl.Collator(locale, {
```
Comment (crobinson42) [inline src/Query.js]: Will this throw if `Intl` is not defined in an environment?

Comment (wickstargazer) [reply src/Query.js]: Ok i will check for undefined and add it the default behavior, that would help?

Comment (crobinson42) [reply src/Query.js]: Yeah, definitely need that. This won't run in Node.js env at all

**Comment by wickstargazer** [comment]
Or is it better to add this Implementation 

https://www.npmjs.com/package/intl-collator

**Comment by crobinson42** [comment]
If we can avoid adding new pkg dependencies that is preferable.

**Comment by wickstargazer** [comment]
Ok then i would add to the docs incase anyone wants to use the feature install extra package in nodejs environment:)

**Comment by crobinson42** [comment]
https://www.js-data.io/v4.0 suggest edits. I would say they belong under queries around `orderBy` syntax

File: src/Query.js (hunk 1 above)
Comment (zuzusik) [reply src/Query.js]: @crobinson42 looks like `Intl` API is available in `Node.js` after all - I've checked it starting from version `8.11.1` (oldest one on my machine) and it works just fine. I am using `Node` installed by `nvm` on `macOS`.

Comment (crobinson42) [reply src/Query.js]: I suppose it wouldn't hurt to only support Node >= 8 in v4 release.. I notice a lot of popular npm packages are dropping support for old Node.js versions. What do you guys think?

Comment (crobinson42) [reply src/Query.js]: We will still need either a polyfill or a graceful failover for `Intl` api: https://caniuse.com/#search=Intl

**Comment by wickstargazer** [comment]
I was in the sense of it being supported in node as well when googling but havent tested it out, thankz @zuzusik :) i will implement a default failover and we can add the support into the documentation i see safari still wont support it?

File: src/Query.js (hunk 1 above)
Comment (zuzusik) [reply src/Query.js]: I think it's a good idea to drop Node below 8.

As for `Intl`, actually `Collator.compare` part of this API is pretty much widely supported: https://caniuse.com/#feat=mdn-javascript_builtins_intl_collator_compare

I wouldn't say we need polyfill, unless we want to support IE 10 and below (which I would drop as well)

**Comment by zuzusik** [comment]
@crobinson42 good call to check https://github.com/js-data/js-data/issues/540 with this one 👍  - it turns out #540 actually gets fixed by this - `Collator.compare` uses same logic as `Array.sort` regarding `null` and `undefined`, putting them at the very end of array in `asc` mode, which is exactly what required in #540.

I will just add appropriate tests tomorrow - will open a PR to `wickstargazer:feature/sort-by-locality`

File: src/Query.js (hunk 1 above)
Comment (wickstargazer) [reply src/Query.js]: I agree that there is no need to support IE 10 if newer browser works, i already added support for edge by force setting locale...
Other than that i can add to the docs and help write a test for the #540 and call it a new day?

Comment (wickstargazer) [reply src/Query.js]: I agree that there is no need to support IE 10 if newer browser works, i already added support for edge by force setting locale...
Other than that i can add to the docs and help write a test for the #540 and call it a new day?

**Comment by zuzusik** [comment]
@wickstargazer btw, I think this should be pointing to `v4` branch, not `master`

**Comment by zuzusik** [comment]
@crobinson42 what do you think - should this be released as `3.x` and then ported to `v4` or should it go to `v4` directly?

**Comment by zuzusik** [comment]
@wickstargazer I have opened a PR in your fork to add tests for #540 https://github.com/wickstargazer/js-data/pull/1

**Comment by wickstargazer** [comment]
would love for it to be in 3.x as its running on our production and 4 is still a distance away...will also merge it into 4

File: src/Query.js (hunk 1 above)
Comment (crobinson42) [reply src/Query.js]: @wick Sounds good 👍

**Comment by crobinson42** [comment]
We can't release in v3 unless there is fallback for non supporting environments because that's a breaking change. v3 supports ie11 and node 0.12 and up... So, if we're going to release for v3 then that requires graceful failing.

**Comment by wickstargazer** [comment]
Hey guys, not an issue if want to put it for v4 only. Or even if i should have a fallback for v3 .. I will be doing some stuff tonight .. 
@zuzusik  the default is "en" for the timebeing IF the browser is Edge. We tested on production, it does not break on other browsers except Edge if locale is null.

File: src/Query.js (hunk 1 above)
Comment (wickstargazer) [reply src/Query.js]: updated the doc, not sure where to put the available since version :)

**Comment by wickstargazer** [comment]
Ok guys, i have changed it to be from v4 and optimized the code a little, also add locale to the reserved keyword (which i am not sure should i do)

@zuzusik  for the Int.Collator i added a default fallback to 'en' Although its mentioned here 
[Intl.Collator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_negotiation)

> the "best fit" matcher lets the runtime provide a locale that's at least, but possibly more, suited for the request than the result of the Lookup algorithm. If the application doesn't provide a locales argument, or the runtime doesn't have a locale that matches the request, then the runtime's default locale is used

So i am guessing although default is provided and locale is not passed it, the sort should work by locality anyways.

**Comment by crobinson42** [comment]
@wickstargazer 201 file changes?

**Comment by wickstargazer** [comment]
i think its because i rebase the forked branch to the original v4 and push force
maybe i should re-merge, let me do it from another branch ....

**Comment by crobinson42** [comment]
@wickstargazer Thanks for this effort! You mentioned you _need_ this in v3 - if you want to PR this feature to v3 and make it optional I'll work with you to get a release out 👍🏼

**Comment by wickstargazer** [comment]
i think we can live until the v4 comes out, we are going to do a small restrcuture in april .. can wait till then!

**Pull request merged**